go/tools/gazelle/resolve/index.go

```diff
@@ -0,0 +1,358 @@
+		// attribute to find the sources. These rules are not automatically
+		// importable from Go.
+
+	case "proto_library":
+		record.lang = config.ProtoLang
+		for _, s := range findSources(rule, buildRel, ".proto") {
+			record.importedAs = append(record.importedAs, importSpec{lang: config.ProtoLang, imp: s})
+		}
+
+	default:
+		return
+	}
+
+	visExpr := rule.Attr("visibility")
+	if visExpr != nil {
+		record.visibility = parseVisibility(visExpr, buildRel)
+	} else {
+		record.visibility = defaultVisibility
+	}
+
+	ix.rules = append(ix.rules, record)
+	ix.labelMap[record.label] = record
+}
+
+// Finish constructs the import index and performs any other necessary indexing
+// actions after all rules have been added. This step is necessary because
+// some rules that are added may later be replaced (existing rules may be
+// replaced by generated rules). Also, for proto rules, we need to be able
+// to dereference a label to find the sources.
+//
+// This function must be called after all AddRulesFromFile and AddGeneratedRules
+// calls but before any findRuleByImport calls.
+func (ix *RuleIndex) Finish() {
+	ix.importMap = make(map[importSpec][]*ruleRecord)
+	oldRules := ix.rules
+	ix.rules = nil
+	for _, r := range oldRules {
+		if r.replaced {
+			continue
+		}
+		rule := bf.Rule{Call: r.call}
+		kind := rule.Kind()
+		if kind == "go_proto_library" || kind == "go_grpc_library" {
+			r.importedAs = findGoProtoSources(ix, r)
+		}
+		for _, imp := range r.importedAs {
+			ix.importMap[imp] = append(ix.importMap[imp], r)
+		}
+	}
+}
+
+type ruleNotFoundError struct {
+	imp     string
+	fromRel string
+}
+
+func (e ruleNotFoundError) Error() string {
+	return fmt.Sprintf("no rule found for %q visible from %s", e.imp, e.fromRel)
+}
+
+func (ix *RuleIndex) findRuleByLabel(label Label, fromRel string) (*ruleRecord, error) {
+	label = label.Abs("", fromRel)
+	r, ok := ix.labelMap[label]
+	if !ok {
+		return nil, ruleNotFoundError{label.String(), fromRel}
+	}
+	return r, nil
+}
+
+// findRuleByImport attempts to resolve an import string to a rule record.
+// imp is the import to resolve (which includes the target language). lang is
+// the language of the rule with the dependency (for example, in
+// go_proto_library, imp will have ProtoLang and lang will be GoLang).
+// fromRel is the slash-separated path to the directory containing the import,
+// relative to the repository root.
+//
+// Any number of rules may provide the same import. If no rules provide
+// the import, ruleNotFoundError is returned. If multiple rules provide the
+// import, this function will attempt to choose one based on visibility.
+// An error is returned if the import is still ambiguous.
+//
+// Note that a rule may be returned even if visibility restrictions will be
+// be violated. Bazel will give a descriptive error message when a build
+// is attempted.
+func (ix *RuleIndex) findRuleByImport(imp importSpec, lang config.Language, fromRel string) (*ruleRecord, error) {
```

Comment:
In the presence of multiple vendor folders, we may need to be able to pick the "nearest" in the hierarchy here as well, but it's not clear to me that there is enough information to do so in the structures.

Reply:
I've added this logic for Go only (not proto). I'm using the Pkg part of the label as a proxy for the directory in the repository. This won't work in flat mode or with manually written rules, but I'm not sure how much vendoring matters for those cases.